src/utils.cpp

```diff
@@ -521,13 +566,69 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
     candidate_paths = ::expand_wildcard_paths(candidate_paths);
 
   std::vector<std::string> valid_executable_paths;
-  for (const auto& path : candidate_paths)
-    if (access(path.c_str(), X_OK) == 0)
-      valid_executable_paths.push_back(path);
+  for (const auto &path : candidate_paths) {
+    std::string rel_path;
+    if (pid > 0 && pid_in_different_mountns(pid))
+      rel_path = path_for_pid_mountns(pid, path);
+    else
+      rel_path = path;
+    if (access(rel_path.c_str(), X_OK) == 0)
+      valid_executable_paths.push_back(rel_path);
+  }
 
   return valid_executable_paths;
 }
 
+std::string path_for_pid_mountns(int pid, const std::string &path) {
+  std::ostringstream pid_relative_path;
+  std::string root =
+      (path.length() >= 1 && path.at(0) == '/') ? "/root" : "/root/";
+  pid_relative_path << "/proc/" << pid << root << path;
+  return pid_relative_path.str();
+}
+
+/*
+Determines if the target process is in the same mount namespace as bpftrace.
+
+If a process is in a different mount namespace (eg, container) it is very
+likely that any references to local paths will not be valid, and that paths
+need to be made relative to the PID.
+
+If an invalid PID is specified or doesn't exist, it returns false.
+True is only returned if the namespace of the target process could be read and
+it doesn't match that of bpftrace.
+*/
+static bool pid_in_different_mountns(int pid) {
+
+  const size_t BUFSIZE = 4096;
```

Comment:
I agonized over this one. Realistically I don't expect this to ever exceed ever 32 bytes, as it is based on a formatted print of a known size, bounded by the the maximum PID size, which is currently 9 digits (4 million is max pid).

Likewise, reading the symlink should always return the strrep of an inode ID for the mount ns, again less than 32 bytes.

However, theoretically as these pare paths, i went with PATH_MAX. It's huge overkill, but probably still so fast to stack allocate that it doesn't matter.

Reply:
thanks for weighing in, i'll stick with 64.